### PR TITLE
OOP-principles / it-academy.by / M-JC1-66-20 / Anton Khramau / Done 'Cat's home' and 'Battle Ground'

### PR DESCRIPTION
**Pull request by Igrok023451**
## Pull request state

Actual status for PR:
- [x] implemented
- [x] reviewed
- [x] fixed
- [ ] approved

### CODE REVIEW

File: src/main/java/com/rakovets/course/java/core/practice/oop/principles/Enemy.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+            this.health = 0;
+        }
+    }
+
+    @Override
+    public boolean isAlive() {
+        if (this.health > 0) {
+            return true;
+        }
+        return false;
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop/principles/Enemy.java:31]: You need to make it easier.

File: src/main/java/com/rakovets/course/java/core/practice/oop/principles/Enemy.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    public void attackHero(Hero hero) {
+        System.out.println("Враг атакует героя");
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop/principles/Enemy.java:35]: You couldn't use `System.out.println` in production code. Please, remove it.

File: src/main/java/com/rakovets/course/java/core/practice/oop/principles/Ghost.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    public Ghost(int health) {
+        super(health);
+    }
+
+    @Override
+    public void attackHero(Hero hero) {
+        System.out.println("Враг атакует героя");
+        if (!(hero instanceof Mag)) {
+            hero.takeDamageHero(15);
+        }
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop/principles/Ghost.java:13]: Do you add `instanceof` for each hero?)
Use abstractions.

File: src/main/java/com/rakovets/course/java/core/practice/oop/principles/cat/Cat.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        return "Mew";
+    }
+
+    public String purr() {
+        return "Purr";
+    }
+
+    public String mew(Person user) {
+        user.changeHappiness(-10);
+        return "Mew";
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop/principles/cat/Cat.java:20]: You can reuse method `mew()`.

File: src/main/java/com/rakovets/course/java/core/practice/oop/principles/cat/Persian.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+    @Override
+    public String mew(Person user) {
+        user.changeHappiness(-15);
+        return "Mew-mew-mew";
+    }
+
+    @Override
+    public String purr(Person user) {
+        user.changeHappiness(15);
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop/principles/cat/Persian.java:26]: What is `15`? Why is `15`?

File: src/test/java/com/rakovets/course/java/core/practice/oop/principles/BattleGroundTest.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    @MethodSource("getPlay")
+    void startBattleGroundTest(int[] expected, Hero hero, Enemy enemy) {
+        for (int i = 0; i < 3; i++) {
+            hero.attackEnemy(enemy);
+            System.out.printf("Здоровье врага: %d\n", enemy.getHealth());
+            enemy.attackHero(hero);
+            System.out.printf("Здоровье героя: %d\n", hero.getHealthHero());
+        }
+        Assertions.assertArrayEquals(expected, new int[]{hero.getHealthHero(), enemy.getHealth()});
+    }
```
Comment (rakovets) [inline src/test/java/com/rakovets/course/java/core/practice/oop/principles/BattleGroundTest.java:30]: Why do you test few cases in ONE test?

File: src/test/java/com/rakovets/course/java/core/practice/oop/principles/CatTest.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.stream.Stream;
+
+class CatTest {
+
+    @Test
+    void mew() {
+        Cat tom = new Cat("Tom");
+        String expectedMew = "Mew";
+        String actualMew = tom.mew();
+        Assertions.assertEquals(expectedMew, actualMew);
```
Comment (rakovets) [inline src/test/java/com/rakovets/course/java/core/practice/oop/principles/CatTest.java:20]: You could separate:
* GIVEN
* WHEN
* THEN

**Review by rakovets — CHANGES_REQUESTED** [review]
Good job!
You need to fix following problems.

File: src/test/java/com/rakovets/course/java/core/practice/oop/principles/CatTest.java (hunk 7 above)
Comment (Igrok023451) [reply src/test/java/com/rakovets/course/java/core/practice/oop/principles/CatTest.java:20]: Why do we need extra spaces?

**Review by rakovets — APPROVED** [review]

**Pull request closed without merging**